File: python/40.组合总和II.py
#
# @lc app=leetcode id=40 lang=python3
#
# [40] Combination Sum II
#
from typing import List

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res  = []
        candidates.sort()
        def dfs(nums, target, path):
            if target < 0: return 
            if target == 0:
                res.append(path)
                return
            for i in range(len(nums)):
                if i>0 and nums[i] == nums[i-1]: continue
                if nums[i] > target: break
                dfs(nums[i+1:], target-nums[i], path+[nums[i]])
        dfs(candidates, target, [])
        
        return res

# if __name__ == "__main__":
#     s = Solution()
#     print(s.combinationSum2([10,1,2,7,6,1,5], 8))

File: cpp/40.组合总和II.cpp
class Solution {
public:
  vector<vector<int>> combinationSum2(vector<int> &candidates, int target) {
    vector<int> cur_res;
    sort(candidates.begin(), candidates.end());
    search(candidates, 0, candidates.size(), cur_res, target);
    return res;
  }

private:
  vector<vector<int>> res;
  void search(vector<int> candidates, int srt, int n, vector<int> cur_res,
              int target) {
    if (target == 0) {
      res.push_back(cur_res);
      return;
    }

    for (int i = srt; i < n && target - candidates[i] >= 0; i++) {
      if (i == srt || candidates[i] != candidates[i - 1]) {
        cur_res.push_back(candidates[i]);
        search(candidates, i + 1, n, cur_res, target - candidates[i]);
        cur_res.pop_back();
      }
    }
  }
};